### PR TITLE
[FIX] point_of_sale : Remove cash for kiosk

### PR DESCRIPTION
**Pull request by SalukiMakingCode**
When changing pos_config settings "self_ordering_mode" to kiosk, verify that there is not pos_payment_methods in cash. Otherwise, when opening the kiosk, it will never be opened because the session is stuck in "opening control".

when pos_config is in kiosk mode, we are not be able to add cash payment_method.

tasks id : 3741234

### CODE REVIEW

File: addons/pos_self_order/models/res_config_settings.py [hunk 1]
```diff
@@ -69,6 +70,11 @@ def _onchange_pos_self_order_pay_after(self):
         if self.pos_self_ordering_pay_after == "each" and not self.module_pos_preparation_display:
             self.module_pos_preparation_display = True
 
+    def _kill_cash_payment(self):
+        for payment_mode in self.pos_payment_method_ids:
+            if payment_mode.is_cash_count:
+                self.pos_payment_method_ids = [(3, payment_mode.id)]
+
```
Comment (vlst-odoo) [inline addons/pos_self_order/models/res_config_settings.py:77]: This is not the way in which we prevent a wrong configuration, rather, we use `[@api.constrains](https://www.odoo.com/documentation/master/developer/reference/backend/orm.html?highlight=api%20constrains#odoo.api.constrains)`

Comment (vlst-odoo) [reply addons/pos_self_order/models/res_config_settings.py:77]: Also, the fix should probably be firstly in `pos.config`. Perhaps you would need some check in `res.config.settings` too, but the main thing should be in `pos.config`

**Review by vlst-odoo — CHANGES_REQUESTED** [review]

File: addons/pos_self_order/models/res_config_settings.py (hunk 1 above)
Comment (SalukiMakingCode) [reply addons/pos_self_order/models/res_config_settings.py:77]: > This is not the way in which we prevent a wrong configuration, rather, we use [`@api.constrains`](https://www.odoo.com/documentation/master/developer/reference/backend/orm.html?highlight=api%20constrains#odoo.api.constrains)

Why do we prefer to frustrate the user rather than help him ?

Comment (davidmonnom) [reply addons/pos_self_order/models/res_config_settings.py:77]: I agree with that, but in `res.config.settings` you should keep onchanges rules. It allow to us to send feedback and interaction to the user before records saving.

``
    @api.onchange("pos_self_ordering_mode", "pos_module_pos_restaurant", "pos_payment_method_ids")
    def _onchange_pos_self_order_kiosk(self):
        if self.pos_self_ordering_mode == 'kiosk':
            self.is_kiosk_mode = True
            self.pos_module_pos_restaurant = False
            self.pos_self_ordering_pay_after = "each"

            for pm in self.pos_payment_method_ids.filtered(lambda x: x.is_cash_count):
                self.pos_payment_method_ids = [(3, pm.id)]
        else:
            self.is_kiosk_mode = False

            if not self.pos_module_pos_restaurant:
                self.pos_self_ordering_service_mode = 'counter'
``

The code above will unlink cash payment methods directly after a user change. BTW you don't need to have a specifics methods for that, you can put the behavior directly in `_onchange_pos_self_order_kiosk`.

``
    @api.onchange("pos_payment_method_ids")
    def _onchange_pos_payment_method_ids(self):
        if self.pos_self_ordering_mode == 'kiosk':
            for pm in self.pos_payment_method_ids:
                if pm.is_cash_count:
                    raise ValidationError(_("You cannot add payment methods in kiosk mode."))
``

Don't forget the second part, it prevent the user adding cash payment after changing mode to kiosk. And it will open a popup to the user with above informations.

---

As Vlad said, you can add a constrains API to the pos_config, which will check the rules when saving. Unlike onchange, which sends feedback to the user directly.

Basically it should be something like thats:

``
    @api.constrains("payment_method_ids")
    def _onchange_payment_method_ids(self):
        for record in self:
            if record.self_ordering_mode == 'kiosk':
                for pm in record.payment_method_ids:
                    if pm.is_cash_count:
                        raise ValidationError(_("You cannot add payment methods in kiosk mode."))
``

Don't forget that validation error should always raise a `ValidationError` in case of error.

[ORM DOC](https://www.odoo.com/documentation/17.0/developer/reference/backend/orm.html)

File: addons/pos_self_order/models/pos_config.py [hunk 2]
```diff
@@ -424,6 +424,8 @@ def action_open_wizard(self):
         self.ensure_one()
 
         if not self.current_session_id:
+            pos_session = self.env['pos.session'].create({'user_id': self.env.uid, 'config_id': self.id})
+            pos_session._ensure_access_token()
```
Comment (davidmonnom) [inline addons/pos_self_order/models/pos_config.py:428]: Nice one we need that, we should merge asap

Comment (vlst-odoo) [reply addons/pos_self_order/models/pos_config.py:428]: we merged this one in a diff pr

File: addons/pos_self_order/models/res_config_settings.py (hunk 1 above)
Comment (vlst-odoo) [reply addons/pos_self_order/models/res_config_settings.py]: I would not use `unlink` in an `onchange` 
image

Comment (vlst-odoo) [reply addons/pos_self_order/models/res_config_settings.py]: also, when a user selects `cash` payment method, you don't want to just remove it. The user has put that method there for a reason. Perhaps they though that it is possible to have `cash` payments. That's why raising a `UserError` in a `constrains` is better UX. You have to inform the user that that option is not possible, perhaps even give some advice on how to proceed in that case. ( for ex by suggestion the paying at counter method )

File: addons/pos_self_order/models/pos_config.py [hunk 3]
```diff
@@ -169,6 +169,14 @@ def _check_default_user(self):
             if record.self_ordering_mode == 'mobile' and not record.self_ordering_default_user_id.has_group("point_of_sale.group_pos_user") and not record.self_ordering_default_user_id.has_group("point_of_sale.group_pos_manager"):
                 raise UserError(_("The Self-Order default user must be a POS user"))
 
+    @api.constrains("payment_method_ids")
+    def _onchange_payment_method_ids(self):
+        for record in self:
+            if record.self_ordering_mode == 'kiosk':
+                for pm in record.payment_method_ids:
+                    if pm.is_cash_count:
+                        raise ValidationError(_("You cannot add cash payment methods in kiosk mode."))
```
Comment (vlst-odoo) [inline addons/pos_self_order/models/pos_config.py:178]: try to avoid this having as much indentation. 
Can you relace this with a generator expression?
Something along the lines of ```
>>> a = [[1,2,3], [4,5,6], [7,8,9]]
>>> any(num == 6 for l in a for num in l)
True
```

File: addons/pos_self_order/models/res_config_settings.py [hunk 4]
```diff
@@ -52,12 +52,21 @@ def _onchange_pos_self_order_kiosk(self):
...
             if not self.pos_module_pos_restaurant:
                 self.pos_self_ordering_service_mode = 'counter'
 
+    @api.onchange("pos_payment_method_ids")
+    def _onchange_pos_payment_method_ids(self):
+        if self.pos_self_ordering_mode == 'kiosk':
+            for pm in self.pos_payment_method_ids:
+                if pm.is_cash_count:
+                    raise ValidationError(_("You cannot add cash payment methods in kiosk mode."))
+
```
Comment (vlst-odoo) [inline addons/pos_self_order/models/res_config_settings.py:69]: i am not entirely convinced that this is needed, but, if you want to keep it, please find a way to not duplicate the code from `pos.config`

File: addons/pos_self_order/models/res_config_settings.py [hunk 5]
```diff
@@ -52,12 +52,21 @@ def _onchange_pos_self_order_kiosk(self):
             self.is_kiosk_mode = True
             self.pos_module_pos_restaurant = False
             self.pos_self_ordering_pay_after = "each"
+            for pm in self.pos_payment_method_ids.filtered(lambda x: x.is_cash_count):
+                self.pos_payment_method_ids = [(3, pm.id)]
```
Comment (vlst-odoo) [inline addons/pos_self_order/models/res_config_settings.py:56]: do not use `unlink` in an `onchange` https://www.odoo.com/documentation/17.0/developer/reference/backend/orm.html#module-odoo.api:~:text=will%20be%20ignored-,Danger,-Since

**Review by vlst-odoo — CHANGES_REQUESTED** [review]

File: addons/pos_self_order/models/res_config_settings.py (hunk 1 above)
Comment (davidmonnom) [reply addons/pos_self_order/models/res_config_settings.py]: Yes, but I don't think it's a big deal, as many onchanges mutate records. For example, when you change Service at, pay after will potentially be changed.

And when a cash payment method is selected in kiosk mode, the validation error (ValidationError) will be lifted directly. The payment method will not disappear directly without information.

I think it's more convenient for the user to have the pop-up window directly when changing the payment method than to have it during registration.

Its up to both of you

Comment (vlst-odoo) [reply addons/pos_self_order/models/res_config_settings.py]: the popup is fine, that's not a problem. The problem is the unlinking operation inside an onchange

Comment (davidmonnom) [reply addons/pos_self_order/models/res_config_settings.py]: @SalukiMakingCode 

Can you try with 

`self.pos_payment_method_ids -= pm.id `

File: addons/pos_self_order/models/res_config_settings.py (hunk 5 above)
Comment (davidmonnom) [reply addons/pos_self_order/models/res_config_settings.py:56]: Try
`self.pos_payment_method_ids -= pm.id`

File: addons/pos_self_order/models/res_config_settings.py [hunk 6]
```diff
@@ -52,12 +52,19 @@ def _onchange_pos_self_order_kiosk(self):
...
             self.is_kiosk_mode = False
 
             if not self.pos_module_pos_restaurant:
                 self.pos_self_ordering_service_mode = 'counter'
 
+    @api.onchange("pos_payment_method_ids")
+    def _onchange_pos_payment_method_ids(self):
+        if self.pos_self_ordering_mode == 'kiosk' and any(pm.is_cash_count for pm in self.pos_payment_method_ids):
+            raise ValidationError(_("You cannot add cash payment methods in kiosk mode."))
+
```
Comment (vlst-odoo) [inline addons/pos_self_order/models/res_config_settings.py:67]: what if you override `payment_method_ids` and add a `domain` property such that the `cash` options are not even displayed? Then you won't need this `onchange` at all. Why display the `cash` options in the dropdown if you are not actually allowed to select them?

Comment (davidmonnom) [reply addons/pos_self_order/models/res_config_settings.py:67]: That's another good solution to me but I think it might be strange for users not to see all their payment methods compared to the onchange error. They might not understand why they're not there.

Again up to you @SalukiMakingCode 

Maybe ask @dama-odoo ?

Comment (SalukiMakingCode) [reply addons/pos_self_order/models/res_config_settings.py:67]: I agree that's a ingenious trick. But I also think it might be strange for the users not to see all their payment method.

@dama-odoo what do you think of that ?

Comment (vlst-odoo) [reply addons/pos_self_order/models/res_config_settings.py:67]: I discussed with @dama-odoo and he agreed with the domain approach. Let's try to do that

**Review by vlst-odoo — APPROVED** [review]
robodoo r+
